src/main/resources/assets/computercraft/lua/rom/programs/http/gist.lua

```diff
@@ -0,0 +1,335 @@
+--- gist.lua - Gist client for ComputerCraft
+-- Made by JackMacWindows for CraftOS-PC and CC: Tweaked
+--
+-- @module gist
+
+local expect = require and require("cc.expect").expect or dofile("/rom/modules/main/cc/expect.lua").expect
+
+-- Internal functions
+
+local function getGistFile(data)
+    if not data.truncated then return data.content else
+        local handle = http.get(data.raw_url)
+        if not handle then error("Could not connect to api.github.com.") end
+        if handle.getResponseCode() ~= 200 then
+            handle.close()
+            error("Failed to download file data.")
+        end
+        local d = handle.readAll()
+        handle.close()
+        return d
+    end
+end
+
+local function setTextColor(c) if term.isColor() then term.setTextColor(c) elseif c == colors.white or c == colors.yellow then term.setTextColor(colors.white) else term.setTextColor(colors.lightGray) end end
+
+local function requestAuth(headers, interactive)
+    if settings.get("gist.id") ~= nil then headers.Authorization = "token " .. settings.get("gist.id") elseif interactive then
+        setTextColor(colors.yellow)
+        write("You need to add a Personal Access Token (PAK) to upload Gists. Follow the instructions at ")
+        setTextColor(colors.blue)
+        write("https://tinyurl.com/GitHubPAK")
+        setTextColor(colors.yellow)
+        write(" to generate one. Make sure to check the '")
+        setTextColor(colors.blue)
+        write("gist")
+        setTextColor(colors.yellow)
+        print("' checkbox on step 7 (under 'Select scopes'). Once done, paste it here.")
+        setTextColor(colors.lime)
+        write("PAK: ")
+        setTextColor(colors.white)
+        local pak = read()
+        if pak == nil or pak == "" then error("Invalid PAK, please try again.") end
+        settings.set("gist.id", pak)
+        headers.Authorization = "token " .. pak
+    end
+end
+
+-- User API - this can be loaded with os.loadAPI
+
+-- ID can be either just the gist ID or a gist ID followed by a slash and a file name
+-- * If a file name is specified, retrieves that file
+-- * Otherwise, if there's only one file, retrieves that file
+-- * Otherwise, if there's a file named 'init.lua', retrieves 'init.lua'
+-- * Otherwise, if there's more than one file but only one *.lua file, retrieves the Lua file
+-- * Otherwise, retrieves the first Lua file alphabetically (with a warning)
+-- * Otherwise, fails
+
+--- Retrieves one file from a Gist using the specified ID.
+-- @tparam string id The Gist ID to download from. See above comments for more details.
+-- @tparam[opt] any showMessages Set to any true value to show messages while connecting.
+-- @treturn string|nil The contents of the specified Gist file, or nil on error.
+-- @treturn string|nil The name of the file that was chosen to be downloaded, or nil on error.
+function get(id, showMessages)
```

Comment:
Instead of a `showMessages` boolean, I wonder if a `progress: function(msg:string)` argument would be better. So we'd just have:


```lua
expect(2, progress, "function", "nil")
progress = progress or do_nothing
-- …
progress("Connecting to api.github.com")
```

Reply:
I've added this. `showMessages` was originally intended to be used only by the `gist` program, but I realize this may be useful for other programs. I didn't do this for `put` and `delete` since they do need a boolean value to know whether to prompt for an ID.